ocs_ci/utility/ibmcloud.py

```diff
@@ -203,3 +205,232 @@ def add_deployment_dependencies():
+                out = run_cmd(cmd)
+                logger.info(f"Node restart command output: {out}")
+
+    def attach_volume(self, volume, node):
+        """
+        Attach volume to node on IBM Cloud.
+
+        Args:
+            volume (str): volume id.
+            node (str): worker node id to attach.
+
+        """
+        logger.info(
+            f"attach_volumes:{node[0].get()['metadata']['labels']['failure-domain.beta.kubernetes.io/zone']}"
+        )
+        provider_id = node[0].get()["spec"]["providerID"]
+        cluster_id = provider_id.split("/")[5]
+        worker_id = node[0].get()["metadata"]["labels"][
+            "ibm-cloud.kubernetes.io/worker-id"
+        ]
+
+        logger.info(f"volume is : {volume}")
+
+        cmd = f"ibmcloud is volume {volume} --output json"
+        out = run_cmd(cmd)
+        out = json.loads(out)
+
+        if len(out["volume_attachments"]) == 0:
+            cmd = (
+                f"ibmcloud ks storage attachment  create --cluster {cluster_id} --worker {worker_id}"
+                f"  --volume {volume} --output json"
+            )
+            out = run_cmd(cmd)
+            out = json.loads(out)
+            logger.info(f"attachment command output: {out}")
+        else:
+            logger.info(f"volume is already attached to node: {out}")
+
+    def detach_volume(self, volume, node=None, delete_from_backend=True):
+        """
+        Detach volume to node on IBM Cloud.
+
+        Args:
+            volume (str): volume id.
+            node (str): worker node id to detach.
+            delete_from_backend (bool): True for deleting the disk from the
+                storage backend, False otherwise
+
+        """
+        provider_id = node[0].get()["spec"]["providerID"]
+        cluster_id = provider_id.split("/")[5]
+        worker_id = node[0].get()["metadata"]["labels"][
+            "ibm-cloud.kubernetes.io/worker-id"
+        ]
+
+        logger.info(f"volume is : {volume}")
+
+        cmd = f"ibmcloud is volume {volume} --output json"
+        out = run_cmd(cmd)
+        out = json.loads(out)
+
+        if out["status"] == "available":
+            attachment_id = out["volume_attachments"][0]["id"]
+
+            cmd = (
+                f"ibmcloud ks storage attachment  rm --cluster {cluster_id} --worker {worker_id}"
+                f"  --attachment {attachment_id}"
+            )
+            out = run_cmd(cmd)
+            logger.info(f"detachment command output: {out}")
+
+    def get_node_by_attached_volume(self, volume):
+        """
+        Get the node by attached volume on IBM Cloud.
+
+        Args:
+            volume (str): volume id.
+
+        """
+        cmd = f"ibmcloud is volume {volume} --output json"
+        out = run_cmd(cmd)
+        out = json.loads(out)
+
+        if not out["volume_attachments"]:
+            logger.info("volume is not attached to node")
+            raise NodeHasNoAttachedVolume("not attached to node")
```

Comment:
The exception name and the reason seems to be slightly different, which might be confusing. The exception name says, that a node have no volume attached, but the reason string (and also the real cause from this method) is that some volume is not attached to any node.

Reply:
changed